go-controller/pkg/ovn/namespace.go

```diff
@@ -240,6 +245,41 @@ func (oc *Controller) AddNamespace(ns *kapi.Namespace) {
 
 	// If multicast enabled, adds all current pods in the namespace to the allow policy
 	oc.multicastUpdateNamespace(ns, nsInfo)
+
+	// ensure any missing network policies are added
+	missingPolicies := make(map[*v1.NetworkPolicy]*networkPolicy)
+
+	for _, policy := range netPolicies {
+		if _, alreadyExists := nsInfo.networkPolicies[policy.Name]; !alreadyExists {
+			if len(nsInfo.networkPolicies) == 0 {
+				if err := oc.ensureDefaultDenyPortGroup(nsInfo, ns.Name, policy.Name); err != nil {
+					klog.Error(err.Error())
+					nsInfo.Unlock()
+					return
+				}
+			}
+
+			np := NewNetworkPolicy(policy)
+			// lock the np to prevent a update/delete event in policy handler from getting it
+			// as we must create it first
+			np.Lock()
+			missingPolicies[policy] = np
+			nsInfo.networkPolicies[policy.Name] = np
+		}
+	}
+
+	aclLogDeny := nsInfo.aclLogging.Deny
+	aclLogAllow := nsInfo.aclLogging.Allow
+	portGroupIngressDenyName := nsInfo.portGroupIngressDenyName
+	portGroupEgressDenyName := nsInfo.portGroupEgressDenyName
+
+	nsInfo.Unlock()
+	for kPolicy, policy := range missingPolicies {
+		// this needs to be in a goroutine to avoid blocking the namespace queue so that if a policy is created
+		// and it adds a namespace handler, it will not be deadlocked due to the add function here not returning
+		// note, we are holding np locks here so no other handler can delete this until its done creating
+		go oc.createNetworkPolicy(policy, kPolicy, aclLogDeny, aclLogAllow, portGroupIngressDenyName, portGroupEgressDenyName)
```

Comment:
there could be a race still I think, because in creating network policy we unlock np, then relock it a few times. During that time a network policy delete event could come, and race with the create. This wasn't a problem with the network policy handler cause the events would come in order there.

Reply:
I realized I'm going down a bad path here. I think it would be better to just leave no creation to the np handler and get the namespace from the lister and stop using nsinfo.

Reply:
Agreed... If you're going to start using the the lister as the source of truth for things then commit to it right? In other places in the code we play the game of maintaining precedence (i.e if NP add event comes before NS ADD create the NS in the NP handler, but doing this seems even safer.. Forget knowing if we created the NS yet in OVN, if it's in the lister we know it either has been/ or will be created so proceed.  

Also, I think holding the np lock for all of createNetworkPolicy is dangerous.  Why not just create the NP there, and then in `addNetworkPolicy` make sure 1. The NS still exists && 2. The policy hasn't changed while you were creating it (via the lister).....then make it.  Then things can start looking more level triggered.

Reply:
You also race on updates - the policy could change in the mean time. So now you could find yourself with an old version of the policy applied.